File: frontend/tests/registri-modul.test.js
'use strict'

// Registri route_key → layar (0.9.33). Menu Beranda desktop & Android disusun dari fungsi murni yang
// SAMA atas manifest server. Semua route_key yang dapat dikirim server (config/pos_archetypes.php +
// pos_verticals.php) harus berujung ke layar, dan setiap arketipe punya pintu transaksi.

const test = require('node:test')
const assert = require('node:assert/strict')

let R

test.before(async () => {
  R = await import('../src/renderer/js/lib/registri-modul.js')
})

const menu = (ids) => ids.map((id, i) => ({ id, routeKey: id, label: '', order: i + 1 }))

// Menu OWNER per arketipe server (route_key, urutan server).
const ARKETIPE = {
  inventory_sale: ['dashboard', 'kasir', 'produk', 'inventory', 'riwayat', 'sesi', 'pelanggan', 'pengeluaran', 'laporan', 'pengaturan'],
  fnb_kot: ['dashboard', 'kasir', 'antrian', 'produk', 'inventory', 'riwayat', 'sesi', 'pelanggan', 'pengeluaran', 'laporan', 'pengaturan'],
  service_lifecycle: ['dashboard', 'order', 'antrian', 'layanan', 'inventory', 'riwayat', 'sesi', 'pelanggan', 'pengeluaran', 'laporan', 'pengaturan'],
  membership: ['dashboard', 'kasir', 'member', 'jadwal', 'riwayat', 'sesi', 'pelanggan', 'pengeluaran', 'laporan', 'pengaturan'],
  hospitality: ['dashboard', 'kamar', 'reservasi', 'riwayat', 'sesi', 'pelanggan', 'pengeluaran', 'laporan', 'pengaturan']
}

test('setiap route_key server (kecuali yang tercatat belum ada) punya layar', () => {
  const semua = new Set(Object.values(ARKETIPE).flat())
  for (const key of semua) {
    if (key === 'dashboard') continue
    assert.ok(R.MODULES[key] && R.MODULES[key].screen, `route_key tanpa layar: ${key}`)
  }
})

test('setiap arketipe punya kartu transaksi (layar kasir) di Beranda', () => {
  for (const [nama, ids] of Object.entries(ARKETIPE)) {
    const kartu = R.susunModul({ menus: menu(ids), manajemen: true })
    assert.ok(kartu.some((k) => k.screen === 'pos'), `${nama}: tidak ada pintu transaksi`)
  }
})

test('urutan & label mengikuti manifest; route_key tak dikenal dilewati dan dilaporkan', () => {
  const peringatan = []
  const kartu = R.susunModul({
    menus: [
      { id: 'riwayat', routeKey: 'riwayat', label: 'Riwayat Nota', order: 2 },
      { id: 'order', routeKey: 'order', label: 'Order Cucian', order: 1 },
      { id: 'x', routeKey: 'teleportasi', label: 'X', order: 3 }
    ],
    manajemen: false,
    peringatan: (k) => peringatan.push(k)
  })
  assert.deepEqual(kartu.map((k) => [k.id, k.title]), [['order', 'Order Cucian'], ['riwayat', 'Riwayat Nota']])
  assert.deepEqual(peringatan, ['teleportasi'])
})

test('fitur ekstra app hanya untuk manajemen; manifest kosong memakai set inti', () => {
  const ids = (k) => k.map((x) => x.id)
  assert.ok(ids(R.susunModul({ menus: menu(['kasir']), manajemen: true })).includes('keuangan'))
  assert.ok(!ids(R.susunModul({ menus: menu(['kasir']), manajemen: false })).includes('keuangan'))
  assert.deepEqual(ids(R.susunModul({ menus: [], manajemen: false })), ['kasir', 'riwayat', 'sesi', 'produk', 'pelanggan', 'laporan', 'pengaturan'])
})

test('kartu utama = pintu transaksi', () => {
  assert.equal(R.kartuUtama('kasir'), true)
  assert.equal(R.kartuUtama('order'), true)
  assert.equal(R.kartuUtama('riwayat'), false)
})

test('route_key jadwal (gym & klinik) berujung ke layar Jadwal', () => {
  assert.equal(R.MODULES.jadwal.screen, 'jadwal')
  const kartu = R.susunModul({ menus: menu(ARKETIPE.membership), manajemen: false })
  assert.ok(kartu.some((k) => k.id === 'jadwal'), 'kartu Jadwal muncul di Beranda membership')
})

test('dua route_key untuk satu tujuan (membership: member + pelanggan) → satu kartu, yang lebih dulu menang', () => {
  const kartu = R.susunModul({ menus: menu(ARKETIPE.membership), manajemen: false })
  const kePelanggan = kartu.filter((k) => k.screen === 'customers')
  assert.equal(kePelanggan.length, 1, 'hanya satu kartu ke layar pelanggan')
  assert.equal(kePelanggan[0].id, 'member')
  assert.equal(kePelanggan[0].title, 'Member')
  assert.ok(kartu.some((k) => k.id === 'jadwal') && kartu.some((k) => k.id === 'sesi'), 'kartu lain tetap ada')
  // Papan yang berbagi layar tetapi beda mode TIDAK ikut tertelan.
  const papan = R.susunModul({
    menus: [
      { id: 'dapur', routeKey: 'dapur', label: '', order: 1 },
      { id: 'antrian', routeKey: 'antrian', label: '', order: 2 }
    ],
    manajemen: false
  })
  assert.deepEqual(papan.map((k) => k.id), ['dapur', 'antrian'])
})

File: mobile-flutter/test/registri_modul_test.dart
// Registri route_key → rute (padanan registri-modul.js desktop). Menu sekunder
// disusun dari manifest server, bukan dari daftar yang ditanam di aplikasi.

import 'dart:io';

import 'package:flutter_test/flutter_test.dart';
import 'package:tuleh_pos/core/navigation/registri_modul.dart';
import 'package:tuleh_pos/features/toko/domain/entities/toko_manifest.dart';

TokoManifest _man(List<String> keys, {List<String> states = const ['SELESAI']}) => TokoManifest(
  menus: [
    for (var i = 0; i < keys.length; i++)
      ManifestMenu(id: keys[i], label: 'Label ${keys[i]}', routeKey: keys[i], order: i + 1),
  ],
  lifecycleStates: states,
);

void main() {
  test('retail: menu sekunder dari manifest, tujuan utama & riwayat (tab ketiga) dilewati', () {
    final menu = menuLainDariManifest(
      _man(['dashboard', 'kasir', 'produk', 'inventory', 'riwayat', 'sesi', 'pelanggan', 'pengeluaran', 'laporan', 'pengaturan']),
    );
    expect(menu.map((m) => m.rute), [
      '/produk', '/sesi', '/pelanggan', '/pengeluaran', '/pengaturan', '/stok',
    ]);
    expect(menu.first.label, 'Produk', reason: 'label registri, bukan label manifest');
  });

  test('toko bertahap: papan jadi tab ketiga sehingga riwayat turun ke Lainnya', () {
    final menu = menuLainDariManifest(
      _man(['dashboard', 'kasir', 'antrian', 'layanan', 'riwayat', 'sesi', 'pengaturan'], states: ['ANTRIAN', 'SELESAI']),
    );
    expect(menu.map((m) => m.rute), ['/produk', '/riwayat', '/sesi', '/pengaturan', '/stok']);
    expect(menu.any((m) => m.rute == '/pesanan'), isFalse, reason: 'papan sudah jadi tujuan utama');
  });

  test('membership: route_key jadwal & member punya rute', () {
    final menu = menuLainDariManifest(_man(['dashboard', 'kasir', 'member', 'jadwal', 'riwayat', 'laporan']));
    expect(menu.map((m) => m.rute), ['/pelanggan', '/jadwal', '/stok']);
    expect(registriModul['jadwal']!.rute, '/jadwal');
  });

  test('route_key tak dikenal diabaikan & dicatat, bukan menu yang menabrak rute mati', () {
    final dicatat = <String>[];
    final menu = menuLainDariManifest(
      _man(['kasir', 'teleportasi', 'inventory', 'pengaturan']),
      peringatan: dicatat.add,
    );
    expect(menu.map((m) => m.rute), ['/pengaturan', '/stok']);
    expect(dicatat, ['teleportasi', 'inventory']);
  });

  test('manifest tanpa menus (server lama): set inti + fitur ekstra app', () {
    expect(
      menuLainDariManifest(const TokoManifest()).map((m) => m.rute),
      ['/produk', '/pelanggan', '/sesi', '/pengaturan', '/stok'],
    );
    expect(menuLainDariManifest(null).map((m) => m.rute).last, '/stok');
  });

  test('setiap rute registri terdaftar di router (tak ada menu ke halaman mati)', () {
    final sumber = File('lib/core/router/app_router.dart').readAsStringSync();
    for (final m in [...registriModul.values.map((e) => e.rute), '/stok']) {
      expect(sumber, contains("'$m'"), reason: 'rute $m belum ada di app_router.dart');
    }
  });
}
